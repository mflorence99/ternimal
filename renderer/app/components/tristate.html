<mat-checkbox
  (click)="next()"
  [color]="(value === null) ? 'warn' : 'accent'"
  [disabled]="disabled"
  [indeterminate]="value === null"
  [ngModel]="value"
>
  <ng-content></ng-content>
</mat-checkbox>
