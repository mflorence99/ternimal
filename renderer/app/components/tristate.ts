import { Component } from '@angular/core';
import { ControlValueAccessor } from '@angular/forms';
import { MAT_CHECKBOX_DEFAULT_OPTIONS } from '@angular/material/checkbox';
import { MatCheckboxDefaultOptions } from '@angular/material/checkbox';
import { NG_VALUE_ACCESSOR } from '@angular/forms';

import { forwardRef } from '@angular/core';

// @see https://stackoverflow.com/questions/49296051

@Component({
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => TriStateComponent),
      multi: true,
    },
    { provide: MAT_CHECKBOX_DEFAULT_OPTIONS, useValue: { clickAction: 'noop' } as MatCheckboxDefaultOptions },
  ],
  selector: 'ternimal-tristate',
  templateUrl: 'tristate.html',
  styleUrls: ['tristate.scss'],
})
export class TriStateComponent implements ControlValueAccessor {

  disabled: boolean;
  
  value: any;
  
  private onChange: Function;
  private onTouched: Function;
  private tape = [null, true, false];

  next(): void {
    this.value = this.tape[(this.tape.indexOf(this.value) + 1) % this.tape.length];
    this.onChange(this.value);
    this.onTouched();
  }

  registerOnChange(fn: Function): void {
    this.onChange = fn;
  }

  registerOnTouched(fn: Function): void {
    this.onTouched = fn;
  }

  setDisabledState(disabled: boolean): void {
    this.disabled = disabled;
  }

  writeValue(value: boolean): void {
    this.value = value ?? this.tape[0];
  }

}
