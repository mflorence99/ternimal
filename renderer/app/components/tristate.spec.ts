import { TriStateComponent } from './tristate';

import { prepare } from './component.spec';

import 'jest-extended';

import { TestBed } from '@angular/core/testing';

describe('TriStateComponent', () => {
  let component: TriStateComponent;

  beforeEach(() => {
    prepare();
    const fixture = TestBed.createComponent(TriStateComponent);
    component = fixture.componentInstance;
  });

  test('writeValue / next', () => {
    component.writeValue(null);
    expect(component.value).toBeNull();
    component.registerOnChange((value) => expect(value).toBeTrue());
    component.registerOnTouched((value) => expect(value).toBeTrue());
    component.next();
  });

  test('disabled state', () => {
    expect(component.disabled).toBeFalse();
    component.setDisabledState(true);
    expect(component.disabled).toBeTrue();
  });
});
