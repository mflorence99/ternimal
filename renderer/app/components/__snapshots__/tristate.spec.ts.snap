// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriStateComponent snapshot 1`] = `
<ternimal-tristate
  __ngContext__={[Function LRootView]}
  disabled="false"
  tape={[Function Array]}
  value={[Function Boolean]}
>
  <mat-checkbox
    class="mat-checkbox mat-accent ng-untouched ng-pristine ng-valid"
  >
    <label
      class="mat-checkbox-layout"
      for="mat-checkbox-3-input"
    >
      <span
        class="mat-checkbox-inner-container mat-checkbox-inner-container-no-side-margin"
      >
        <input
          aria-checked="false"
          class="mat-checkbox-input cdk-visually-hidden"
          id="mat-checkbox-3-input"
          tabindex="0"
          type="checkbox"
        />
        <span
          class="mat-ripple mat-checkbox-ripple mat-focus-indicator"
          matripple=""
        >
          <span
            class="mat-ripple-element mat-checkbox-persistent-ripple"
          />
        </span>
        <span
          class="mat-checkbox-frame"
        />
        <span
          class="mat-checkbox-background"
        >
          <svg
            class="mat-checkbox-checkmark"
            focusable="false"
            version="1.1"
            viewBox="0 0 24 24"
            xml:space="preserve"
          >
            <path
              class="mat-checkbox-checkmark-path"
              d="M4.1,12.7 9,17.6 20.3,6.3"
              fill="none"
              stroke="white"
            />
          </svg>
          <span
            class="mat-checkbox-mixedmark"
          />
        </span>
      </span>
      <span
        class="mat-checkbox-label"
      >
        <span
          style="display: none;"
        >
          Â 
        </span>
      </span>
    </label>
  </mat-checkbox>
</ternimal-tristate>
`;
